inductor: add functionalization pass for customer mkldnn_convolution_pointwise_binary_ inplace op

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94172

There always has a runtime waning when running inductor CPU path:
```
l.  Known situations this can occur are inference mode only compilation involving resize_ or prims (!schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED); if your situation looks different please file a bug to PyTorch.
Traceback (most recent call last):
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 1377, in aot_wrapper_dedupe
    fw_metadata, _out = run_functionalized_fw_and_collect_metadata(flat_fn)(
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 578, in inner
    flat_f_outs = f(*flat_f_args)
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 2455, in functional_call
    out = Interpreter(mod).run(*args[params_len:], **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 136, in run
    self.env[node] = self.run_node(node)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 177, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 294, in call_module
    return submod(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 344, in forward
    return self._conv_forward(input, other, self.weight, self.bias)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 327, in _conv_forward
    return torch.ops.mkldnn._convolution_pointwise_(
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
  File "/home/xiaobing/pytorch-offical/torch/_inductor/overrides.py", line 38, in __torch_function__
    return func(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: !schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED at "/home/xiaobing/pytorch-offical/aten/src/ATen/FunctionalizeFallbackKernel.cpp":32, please report a bug to PyTorch. mutating and aliasing ops should all have codegen'd kernels

While executing %self_layer2_0_downsample_0 : [#users=2] = call_module[target=self_layer2_0_downsample_0](args = (%self_layer1_1_conv2, %self_layer2_0_conv2), kwargs = {})
Original traceback:
  File "/home/xiaobing/vision/torchvision/models/resnet.py", line 100, in forward
    identity = self.downsample(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 274, in _forward_impl
    x = self.layer2(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 285, in forward
    return self._forward_impl(x)
```
 
This PT adds a functional pass for the mkldnn_convolution_pointwise_binary_ in-place op to avoid this warning message.

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10

XiaobingSuper committed:
inductor: add functionalization pass for customer mkldnn_convolution_pointwise_binary_ inplace op

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -919,6 +998,14 @@ TORCH_LIBRARY_IMPL(mkldnn, Meta, m) {
 }
+
+
+TORCH_LIBRARY_IMPL(mkldnn, Functionalize, m) {
+  m.impl(
+      TORCH_SELECTIVE_NAME("mkldnn::_convolution_pointwise_.binary"),
+      TORCH_FN(functionalization::mkldnn_convolution_pointwise_binary_));
+}
```
@bdhirsh For inplace ops not registered via native_functions.yaml like this one, do we have an easier way of implementing the functionalization without explicit implementation op-by-op?

bdhirsh replied:
Yes this is very boilerplate-y... Long term, we should write a generic / boxed functionalization kernel that works for arbitrary mutable/aliasing ops. But until then, writing a one-off kernel here seems acceptable.

bdhirsh reviewed (approved):
left a minor question, but stamped to unblock

bdhirsh commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -884,6 +885,84 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_convolution_backward(
+    tmp_output = other_.add(conv_output);
+  }
+
+  at::functionalization::impl::replace_(other_t, conv_output);
```
Shouldn't this be `tmp_output`, not `conv_output`? Otherwise, you're ignoring the results of the (inplace) binary and unary op.

XiaobingSuper replied:
yes,  changed.

XiaobingSuper committed:
Update on "inductor: add functionalization pass for customer mkldnn_convolution_pointwise_binary_ inplace op"


There always has a runtime waning when running inductor CPU path:
```
l.  Known situations this can occur are inference mode only compilation involving resize_ or prims (!schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED); if your situation looks different please file a bug to PyTorch.
Traceback (most recent call last):
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 1377, in aot_wrapper_dedupe
    fw_metadata, _out = run_functionalized_fw_and_collect_metadata(flat_fn)(
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 578, in inner
    flat_f_outs = f(*flat_f_args)
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 2455, in functional_call
    out = Interpreter(mod).run(*args[params_len:], **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 136, in run
    self.env[node] = self.run_node(node)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 177, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 294, in call_module
    return submod(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 344, in forward
    return self._conv_forward(input, other, self.weight, self.bias)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 327, in _conv_forward
    return torch.ops.mkldnn._convolution_pointwise_(
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
  File "/home/xiaobing/pytorch-offical/torch/_inductor/overrides.py", line 38, in __torch_function__
    return func(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: !schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED at "/home/xiaobing/pytorch-offical/aten/src/ATen/FunctionalizeFallbackKernel.cpp":32, please report a bug to PyTorch. mutating and aliasing ops should all have codegen'd kernels

While executing %self_layer2_0_downsample_0 : [#users=2] = call_module[target=self_layer2_0_downsample_0](args = (%self_layer1_1_conv2, %self_layer2_0_conv2), kwargs = {})
Original traceback:
  File "/home/xiaobing/vision/torchvision/models/resnet.py", line 100, in forward
    identity = self.downsample(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 274, in _forward_impl
    x = self.layer2(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 285, in forward
    return self._forward_impl(x)
```
 
This PT adds a functional pass for the mkldnn_convolution_pointwise_binary_ in-place op to avoid this warning message.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -884,6 +885,84 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_convolution_backward(
+  }
+
+  if (unary_attr.has_value()) {
+    tmp_output = other_.add(conv_output).relu();
```
For the functionalization of such an in-place op, we just want to make sure it can work when collecting the meta info, and that it will lower the fused kernel at the inductor path.

bdhirsh replied:
I see. In that case - if you're okay with decomposing the op (and relying on fusion for perf), then I think there's a simpler alternative here, that doesn't require a custom functionalization kernel. You can write a python decomp for this op, that just decomposes it into `convolution` + inplace add + inplace unary op. And then you can rely on functionalization's existing support for the inplace add + unary ops. The easiest way to write this decomp is in python, registering it to the autograd key (so it decomposes before functionalization). Example [here](https://github.com/pytorch/pytorch/blob/6ac0198c0228a5b59f337cac5a752dbb28e341e5/torch/_decomp/decompositions.py#L1435) (although it's a little unclear to me if the decomp should live directly in that file - does that mkldnn operator always exist in the dispatcher? `mkldnn/Conv.cpp` only get built + that op registered on a build with mkldnn).

bdhirsh replied:
> and that it will lower the fused kernel at the inductor path.

I might have misunderstood. This function as written won't end up lowering the fused kernel - It will directly lower `mkldnn_convolution` + `aten.add` + `aten.relu` to inductor

XiaobingSuper replied:
@bdhirsh , for my understanding, the functionalization is used in https://github.com/pytorch/pytorch/blob/6ac0198c0228a5b59f337cac5a752dbb28e341e5/torch/_functorch/aot_autograd.py#L1378 to collect the meta info, but we don't want to do the decomposing, so for inductor side, we still get the in -place fusion op, and lowering same the in-place kernel:https://github.com/pytorch/pytorch/blob/ca63040d2b7583ae7c8f066fff302473b591f194/torch/_inductor/lowering.py#L902.

jgong5 replied:
Yep, this op is not user-generated but a fusion op generated from inside the inductor backend. The reason why we have to explicitly implement a functionalization is that the fusion path is currently added before AOTAutograd. We have the plan to move the fusions after AOTAutograd later so that this functionalization can be removed after the move is completed.

XiaobingSuper replied:
> 

I don't think it will lower tto **mkldnn_convolution** + **aten.add** + **aten.relu** to inductor, for the inductor path, if we don't register a decomposing at https://github.com/pytorch/pytorch/blob/6c80d0a5a55de71d0cc56950f7459b5c334a14ab/torch/_decomp/__init__.py#L175, we can't call in the decomposition path even we do https://github.com/pytorch/pytorch/blob/6ac0198c0228a5b59f337cac5a752dbb28e341e5/torch/_decomp/decompositions.py, is it right?

bdhirsh replied:
Hmm.. so there are a few ways for ops to get "decomposed" in AOTAutograd. One is through the explicit decomp table, like you said. But separately from the decomp table, both autograd and functionalization run inside of AOTAutograd. If either of them chooses to directly decompose an operator, then the compiler sitting underneath AOTAutograd will see the result of that decomposition too.

You can probably try writing a simple test to confirm whether or not inductor will see the fused op:
```
@torch.compile
def f(inp, other, weight):
    # added a clone so that the mutation is not a "graph input mutation"
    other_clone = other.clone()
    return torch.ops.mkldnn.mkldnn_convolution_pointwise_binary_(inp, other, weight)



inp = torch.ones(64)
other = torch.ones(64)
weight = torch.ones(64, requires_grad=True)
out = f(inp, other, weight)
```


FWIW - AOTAutograd has two code paths: `aot_dispatch_autograd` and `aot_dispatch_base`, and only the first actually runs functionalization - so if you are going the "inference path", then you won't see your functionalization kernel actually affect the graph sent to inductor (this happens if none of the inputs to the graph require grad). But before the 2.0 branch cut, I'm planning on getting functionalization to run in both paths of AOTAutograd.

EikanWang commented:
> left a minor question, but stamped to unblock

@bdhirsh , do you mean we can merge this PR since it is close to PT2.0 branch cut date?

XiaobingSuper committed:
Update on "inductor: add functionalization pass for customer mkldnn_convolution_pointwise_binary_ inplace op"


There always has a runtime waning when running inductor CPU path:
```
l.  Known situations this can occur are inference mode only compilation involving resize_ or prims (!schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED); if your situation looks different please file a bug to PyTorch.
Traceback (most recent call last):
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 1377, in aot_wrapper_dedupe
    fw_metadata, _out = run_functionalized_fw_and_collect_metadata(flat_fn)(
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 578, in inner
    flat_f_outs = f(*flat_f_args)
  File "/home/xiaobing/pytorch-offical/torch/_functorch/aot_autograd.py", line 2455, in functional_call
    out = Interpreter(mod).run(*args[params_len:], **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 136, in run
    self.env[node] = self.run_node(node)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 177, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/home/xiaobing/pytorch-offical/torch/fx/interpreter.py", line 294, in call_module
    return submod(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 344, in forward
    return self._conv_forward(input, other, self.weight, self.bias)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/mkldnn.py", line 327, in _conv_forward
    return torch.ops.mkldnn._convolution_pointwise_(
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
  File "/home/xiaobing/pytorch-offical/torch/_inductor/overrides.py", line 38, in __torch_function__
    return func(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_ops.py", line 499, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: !schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED at "/home/xiaobing/pytorch-offical/aten/src/ATen/FunctionalizeFallbackKernel.cpp":32, please report a bug to PyTorch. mutating and aliasing ops should all have codegen'd kernels

While executing %self_layer2_0_downsample_0 : [#users=2] = call_module[target=self_layer2_0_downsample_0](args = (%self_layer1_1_conv2, %self_layer2_0_conv2), kwargs = {})
Original traceback:
  File "/home/xiaobing/vision/torchvision/models/resnet.py", line 100, in forward
    identity = self.downsample(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 274, in _forward_impl
    x = self.layer2(x)
 |   File "/home/xiaobing/vision/torchvision/models/resnet.py", line 285, in forward
    return self._forward_impl(x)
```
 
This PT adds a functional pass for the mkldnn_convolution_pointwise_binary_ in-place op to avoid this warning message.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

bdhirsh commented:
Landing this PR to unblock that error sounds ok to me (before the release), but I'm still not convinced that your custom op won't get decomposed before going to inductor :)

XiaobingSuper commented:
> Landing this PR to unblock that error sounds ok to me (before the release), but I'm still not convinced that your custom op won't get decomposed before going to inductor :)

Currently, only the inference path can call this custom op, this is why I didn't see the decomposition(we check whether the **torch.is_grad_enabled()**)
https://github.com/pytorch/pytorch/blob/782e4f5c02abaf5b9cdba4eaa827bc70a310bca8/torch/_functorch/aot_autograd.py#L2164-L2166

XiaobingSuper commented:
@bdhirsh, yes, you are right, if applying your PR https://github.com/pytorch/pytorch/pull/92857, the customer op will be get decomposed. Can your PR be merged before the 2.0 branch cut(Feb.10)? Thanks!

bdhirsh commented:
@XiaobingSuper my goal is to get that PR in by the 2.0 release, yep.

bdhirsh commented:
> Another question, do we have a method that implements the customer in-place op's functionalization, but doesn't do the decomposition?

> or don't introduce this in-place op before AOTAutograd, after getting the fx graph, we can replace some outplace ops with in-place ops.


Hmmm let me know if I didn't understand your question. But one way to avoid the decomposition here is to:

(1) introduce a new op, `mkldnn_convolution_pointwise_binary`, that is an out-of-place equivalent of `mkldnn_convolution_pointwise_binary_`.

(2) Update that functionalization kernel to directly call `mkldnn_convolution_pointwise_binary`, instead of manually calling each of the decomposed operators.

jgong5 commented:
We now prefer https://github.com/pytorch/pytorch/pull/94581 to this PR as a way to handle the inplace fusion ops in inductor.
